### PR TITLE
validation for duplicate dag ids

### PR DESCRIPTION
**Pull request by jay-bhambhani**
it was pretty complicated to write tests for this given the need for a dagbag and all the airflow stuff. tested manually by adding lines to dags/tests and ensuring that it worked as expected.

### CODE REVIEW

**Review by jpugliesi — COMMENTED** [review]
An alternative to stripping the characters from the string is to use `eval`. this example parses the dag_ids instead of the counter
```
        # absolute hack to extract each individual file's dag ids
        dags = [dag for stat in dagbag.dagbag_stats for dag in eval(stat.dags)]
        duplicate_dags = set([dag for dag in dags if dags.count(dag) > 1])

        if duplicate_dags:
            print(f'Duplicate dag_ids found: {duplicate_dags}')
            assert False

```

File: .pre-commit-hooks.yaml [hunk 1]
```diff
@@ -25,3 +25,9 @@
   entry: check_notebook_has_parameters_cell
   language: python
   files: (^|/).+\.ipynb$
+
+- id: validate_duplicate_dag_ids
+  name: 'Ensure that we do not have duplicate dag ids names'
+  description: 'Ensure that we do not have duplicate dag ids names'
```
Comment (jpugliesi) [inline .pre-commit-hooks.yaml:31]:
```suggestion
  description: 'Ensure that we do not have duplicate dag ids'
```

File: .pre-commit-hooks.yaml [hunk 2]
```diff
@@ -25,3 +25,9 @@
   entry: check_notebook_has_parameters_cell
   language: python
   files: (^|/).+\.ipynb$
+
+- id: validate_duplicate_dag_ids
+  name: 'Ensure that we do not have duplicate dag ids names'
```
Comment (jpugliesi) [inline .pre-commit-hooks.yaml:30]:
```suggestion
  name: 'Ensure that we do not have duplicate dag ids'
```

**Review by jpugliesi — CHANGES_REQUESTED** [review]
This actually is not very portable because it assumes the runtime environment has python with airflow correctly installed, with dependencies. A better approach imo would be:
1. Add a marked test for this validation (under `test_import_dags.py`:
```
    @pytest.mark.precommit
    @pytest.mark.slow
    def test_duplicate_dags(self, dagbag):
        """Test for duplicate DAGs."""
        # absolute hack to extract each individual file's dag ids
        dags = [dag for stat in dagbag.dagbag_stats for dag in eval(stat.dags)]
        duplicate_dags = set([dag for dag in dags if dags.count(dag) > 1])

        if duplicate_dags:
            print(f'Duplicate dag_ids found: {duplicate_dags}')
            assert False
```
2. Create a make recipe to run this test only:
```
.PHONY: pre-commit
pre-commit:
	# Runs pre-commit tests
	docker build -t viaduct.ai/airflow-test -f tests/Dockerfile ${PWD}
	docker run viaduct.ai/airflow-test '-m precommit'
```

3. Run `make pre-commit` in the pre-commit hook

**Comment by jay-bhambhani** [comment]
> An alternative to stripping the characters from the string is to use `eval`. this example parses the dag_ids instead of the counter
> 
> ```
>         # absolute hack to extract each individual file's dag ids
>         dags = [dag for stat in dagbag.dagbag_stats for dag in eval(stat.dags)]
>         duplicate_dags = set([dag for dag in dags if dags.count(dag) > 1])
> 
>         if duplicate_dags:
>             print(f'Duplicate dag_ids found: {duplicate_dags}')
>             assert False
> ```

the only reason i didn't use eval is that I have always been told it is unsafe - so I try my best to avoid usage where I can.

**Comment by devstein** [comment]
Why do we need this as `pre-commit` hook? Why not only make it part of CI?

**Comment by jay-bhambhani** [comment]
it was asked for as a precommit hook https://viaduct-ai.slack.com/archives/CQ397F9GA/p1614806104054300. we could make it a test, but i think it is quite nice to have as a pre-commit hook

**Comment by jay-bhambhani** [comment]
closing to make ci

**Pull request closed without merging**